src/runtime_src/hip/core/event.cpp

```diff
@@ -171,22 +169,23 @@ bool kernel_start::wait()
   return false;
 }
 
-copy_buffer::copy_buffer(std::shared_ptr<stream> s, xclBOSyncDirection direction, std::shared_ptr<memory> buf, void* ptr, size_t size, size_t offset)
-  : command(std::move(s)), cdirection(direction), buffer(std::move(buf)), host_ptr(ptr), copy_size(size), dev_offset(offset)
+template<class T>
+copy_buffer<T>::copy_buffer(std::shared_ptr<stream> s, xclBOSyncDirection direction, std::shared_ptr<memory> buf, T host_buf, size_t size, size_t offset)
```

Comment:
This looks very complicated without understanding what is happening and why?

Reply:
The original template use is just attempt for incorporate two use case: 1. source/destination host buffer is an opaque handle. 2. host buffer is source only and its storage need to be released by class copy buffer after async copy operation.